File: OC/Components/DYTableView.m
//
//  DYTableView.m
//  ID贷
//
//  Created by apple on 2019/6/20.
//  Copyright © 2019 hansen. All rights reserved.
//

#import "DYTableView.h"
#import "MJRefresh.h"
#import "WDButton.h"

@interface DYTableView ()<UITableViewDelegate,UITableViewDataSource>
@property (nonatomic,strong) WDButton *noDataBtn;


@end
@implementation DYTableView

@synthesize noDataText = _noDataText;
@synthesize noDataImage = _noDataImage;

- (instancetype)initWithFrame:(CGRect)frame style:(UITableViewStyle)style {
    
    self = [super initWithFrame:frame style:style];
    [self setupView];
    
    return self;
}

- (void)awakeFromNib {
    [super awakeFromNib];
    [self setupView];
    
}


- (void)setupView {
    self.isShowNoData = YES;
    self.mj_header = [MJRefreshNormalHeader headerWithRefreshingTarget:self refreshingAction:@selector(loadData)];
    self.tableFooterView = [UIView new];
    self.dy_dataSource = @[].mutableCopy;

    self.dataSource = self;
    self.delegate = self;
    self.pageSize = 10;
    self.pageIndex = 0;
    
    [self addSubview:self.noDataBtn];
    [self.noDataBtn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.center.offset(0);
    }];
    [self.noDataBtn addTarget:self action:@selector(nodataBtnClick) forControlEvents:UIControlEventTouchUpInside];
    [_noDataBtn setTitle:self.noDataText forState:UIControlStateNormal];
    [_noDataBtn setImage:[UIImage imageNamed:self.noDataImage] forState:UIControlStateNormal];
}

- (void)nodataBtnClick {
    
    [self.mj_header beginRefreshing];
    self.noDataBtn.hidden = YES;
    
}



- (void)loadData {
    self.pageIndex = 0;
    
    if (self.loadDataCallback) {
        kWeakly(self);
        self.loadDataCallback(self.pageIndex, ^(NSArray * _Nonnull sources) {
            [weakself handleDragDownData:sources];
        });
    }
    
}

- (void)loadMoreData {
    self.pageIndex += 1;
    if (self.loadDataCallback) {
        kWeakly(self);
        self.loadDataCallback(self.pageIndex, ^(NSArray * _Nonnull sources) {
            [weakself handleUpDragData:sources];
        });
    }
    
}
/**下拉刷新*/

- (void)handleDragDownData:(NSArray *)sources {
    
    [self.mj_header endRefreshing];
    [self.dy_dataSource removeAllObjects];
    [self.dy_dataSource addObjectsFromArray:sources];
    [self reloadData];
    
    if (sources.count > 0 && sources.count == self.pageSize) {
        self.mj_footer = [MJRefreshAutoFooter footerWithRefreshingTarget:self refreshingAction:@selector(loadMoreData)];
    }
    
    if (sources.count < self.pageSize && self.mj_footer) {
        [self.mj_footer endRefreshingWithNoMoreData];
    }
    
    if (sources.count == 0 && self.isShowNoData) {
        self.noDataBtn.hidden = NO;
    }
}

- (void)handleUpDragData:(NSArray *)sources {
    [self.mj_footer endRefreshing];
    NSMutableArray *indexes = [[NSMutableArray alloc] initWithCapacity:sources.count];
    for (id objc in sources) {
        [indexes addObject:[NSIndexPath indexPathForRow:self.dy_dataSource.count inSection:0]];
        [self.dy_dataSource addObject:objc];
        
    }
    [self insertRowsAtIndexPaths:indexes withRowAnimation:0];
    if (sources.count < self.pageSize) {
        [self.mj_footer endRefreshingWithNoMoreData];
    }

}

- (void)setNoDataText:(NSString *)noDataText {
    _noDataText = noDataText;
    [_noDataBtn setTitle:self.noDataText forState:UIControlStateNormal];

}
- (void)setNoDataImage:(NSString *)noDataImage {
    _noDataImage = noDataImage;
    [_noDataBtn setImage:[UIImage imageNamed:self.noDataImage] forState:UIControlStateNormal];

}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    
    return self.dy_dataSource.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    DYTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell" forIndexPath:indexPath];
    cell.model = self.dy_dataSource[indexPath.row];
    
    kWeakly(self);
    cell.OtherClickFlag = ^(NSInteger flag) {
        if (weakself.didSelectedCellCallback) {
            weakself.didSelectedCellCallback(indexPath, @(flag));
        }
    };
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    if (self.didSelectedCellCallback) {
        self.didSelectedCellCallback(indexPath,nil);
    }
    
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    if (self.rowHeight > 0) {
        return self.rowHeight;
    }
    
    CGFloat height = [self.heightCache[@(indexPath.row)] floatValue];
    if (height == 0) {
        height = [self.dy_dataSource[indexPath.row] calculateCellheight];
        self.heightCache[@(indexPath.row)] = @(height);
    }
    return height;
    
}

- (WDButton *)noDataBtn {
    
    if (!_noDataBtn) {
        _noDataBtn = [WDButton new];
        _noDataBtn.direction = 1;
        _noDataBtn.hidden = YES;
        _noDataBtn.titleLabel.font = [UIFont systemFontOfSize:15];
        [_noDataBtn setTitleColor:[UIColor hex:@"#555555"] forState:UIControlStateNormal];
        
    }
    return _noDataBtn;
}

- (NSString *)noDataText {
    if (_noDataText.length == 0) {
        _noDataText = @"暂无数据";
    }
    return _noDataText;
}
- (NSString *)noDataImage {
    if (_noDataImage.length == 0) {
        _noDataImage = @"noDataImage";
    }
    return _noDataImage;
}
/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

@end



@implementation DYTableViewCell



@end

@implementation DYTableViewModel



- (CGFloat)calculateCellheight {
    
    NSAssert(1, @"You must rewrite this method in subclass.");
    return 0;
}
@end

File: Swift/iOS_HSLX/华商领袖/classes/UIComponents/DYTableView.swift
//
//  DYTableView.swift
//  华商领袖
//
//  Created by hansen on 2019/4/28.
//  Copyright © 2019 huashanglingxiu. All rights reserved.
//

import UIKit
import MJRefresh
import Moya



/*暂时只支持单组列表*/
class DYTableView<Request: TargetType, Model: BaseModel>: UITableView {

    
    
    var pageIndex = 0;
    /**
     * 当设置size== 0时 不能上拉加载
     *
     */
    var pageSize = 10;
    ///数据中所对应的数组key
    var listKey: String?
    ///上下拉刷新 获取网络请求 必须
    var loadDataCallback: ((_ pageIndex: Int) -> Request)?
   
    ///翻页时要刷新的那组数据
    var section: Int = 0
    //接口请求成功后的回调原始数据 type 0 = 下拉加载 1 = 上拉加载
    var loadDataFinished: ((_ response : Any, _ type: Int) -> Void)?
    
    var dataSources: [Model] = []
    
    override init(frame: CGRect, style: UITableView.Style) {
        super.init(frame: frame, style: style)
        self.mj_header = MJRefreshNormalHeader.init(refreshingTarget: self, refreshingAction: #selector(loadData))
        self.tableFooterView = UIView.init()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    
    @objc private func loadData() {
        guard let networkType = self.loadDataCallback?(0) else {
            debugPrint("No load request!!!")
            return
        }
        self.pageIndex = 0
        NetWorkRequest(target: networkType) { (result) in
        
            self.mj_header.endRefreshing()
            switch result {
            case .successful(let response):
                
                self.loadDataFinished?(response, 0)
                self.dataSources.removeAll()
                var array: [[String : Any]]? = nil
                if (self.listKey?.count ?? 0) > 0 {
                    //使用这个key取相对应的数据
                    guard let resultArray = (response as! [String : Any])[self.listKey!] as? [[String : Any]] else {
                        return
                    }
                    array = resultArray
                } else {
                    guard let resultArray = response as? [[String : Any]] else {
                        return
                    }
                    array = resultArray
                }
                
                for item in array! {
                    self.dataSources.append(Model.init(JSON: item)!)
                }
                self.reloadData()
                if array!.count > 0  && self.pageSize > 0{
                    self.mj_footer = MJRefreshAutoNormalFooter.init(refreshingTarget: self, refreshingAction: #selector(self.loadMoreData))
                }
                if array!.count < self.pageSize && self.mj_footer !== nil {
                    self.mj_footer.endRefreshingWithNoMoreData()
                }
                break
                
            case .failure(let errmsg) :
                SVProgressHUD.showError(withStatus: errmsg)
                break
                
            }
            
        }
        
    }
    
    @objc private func loadMoreData () {
        self.pageIndex += 1;
        guard let networkType = self.loadDataCallback?(self.pageIndex) else {
            debugPrint("No load more  request!!!")
            return
        }
        NetWorkRequest(target: networkType) { (result) in
            self.mj_footer.endRefreshing()
            switch result {
            case .successful(let response):
                self.loadDataFinished?(response, 1)
                var array: [[String : Any]]? = nil
                if (self.listKey?.count ?? 0) > 0 {
                    //使用这个key取相对应的数据
                    guard let resultArray = (response as! [String : Any])[self.listKey!] as? [[String : Any]] else {
                        return
                    }
                    array = resultArray
                } else {
                    guard let resultArray = response as? [[String : Any]] else {
                        return
                    }
                    array = resultArray
                }
                var indexes: [IndexPath] = []
                for item in array! {
                    indexes.append(IndexPath.init(row: self.dataSources.count, section: self.section))
                    self.dataSources.append(Model.init(JSON: item)!)
                }
                self.insertRows(at: indexes, with: UITableView.RowAnimation.none)
                if array!.count < self.pageSize {
                    self.mj_footer.endRefreshingWithNoMoreData()
                }
                break
            case .failure(let errmsg) :
                
                SVProgressHUD.showError(withStatus: errmsg)
                break
            }
        }
    }
    
    deinit {
        self.loadDataCallback = nil
    }
    
    /*
    // Only override draw() if you perform custom drawing.
    // An empty implementation adversely affects performance during animation.
    override func draw(_ rect: CGRect) {
        // Drawing code
    }
    */

}



